test/contracts/PermitAndCall.test.ts

```diff
@@ -28,15 +29,47 @@ describe('Permitable', function () {
         expect(await erc20PermitMock.allowance(signer1.address, permitAndCallMock.target)).to.equal(0);
     });
+
+    it('should work with nested permit', async function () {
+        const { permitAndCallMock, erc20PermitMock, chainId } = await loadFixture(deployTokens);
+        const permit1 = await getPermit(signer1, erc20PermitMock, '1', chainId, await permitAndCallMock.getAddress(), value.toString(), 0x8fffffff.toString(), true);
+        const permit2 = await getPermit(signer2, erc20PermitMock, '1', chainId, await permitAndCallMock.getAddress(), value.toString(), 0x8fffffff.toString(), false);
+        const fooCall = (await permitAndCallMock.foo.populateTransaction()).data;
+        const innerPermitCalldata = (await permitAndCallMock.permitAndCall.populateTransaction(erc20PermitMock.target + trim0x(permit2), fooCall)).data;
+        const tx = await permitAndCallMock.permitAndCall(erc20PermitMock.target + trim0x(permit1), innerPermitCalldata);
+        await expect(tx).to.emit(permitAndCallMock, 'FooCalled');
+        expect(await erc20PermitMock.allowance(signer1.address, permitAndCallMock.target)).to.equal(value);
+        expect(await erc20PermitMock.allowance(signer2.address, permitAndCallMock.target)).to.equal(value);
+    });
+
+    it('should work with payable function', async function () {
+        const { permitAndCallMock, erc20PermitMock, chainId } = await loadFixture(deployTokens);
+        const permit = await getPermit(signer1, erc20PermitMock, '1', chainId, await permitAndCallMock.getAddress(), value.toString(), 0x8fffffff.toString(), true);
+        const tx = await permitAndCallMock.permitAndCall(erc20PermitMock.target + trim0x(permit), (await permitAndCallMock.payableFoo.populateTransaction()).data, { value: 1n });
+        await expect(tx).to.emit(permitAndCallMock, 'MsgValue').withArgs(1n);
+        expect(await erc20PermitMock.allowance(signer1.address, permitAndCallMock.target)).to.equal(value);
+    });
+
+    it('should work with payable function and nested permit', async function () {
+        const { permitAndCallMock, erc20PermitMock, chainId } = await loadFixture(deployTokens);
+        const permit1 = await getPermit(signer1, erc20PermitMock, '1', chainId, await permitAndCallMock.getAddress(), value.toString(), 0x8fffffff.toString(), true);
+        const permit2 = await getPermit(signer2, erc20PermitMock, '1', chainId, await permitAndCallMock.getAddress(), value.toString(), 0x8fffffff.toString(), false);
+        const fooCall = (await permitAndCallMock.payableFoo.populateTransaction()).data;
+        const innerPermitCalldata = (await permitAndCallMock.permitAndCall.populateTransaction(erc20PermitMock.target + trim0x(permit2), fooCall)).data;
+        const tx = await permitAndCallMock.permitAndCall(erc20PermitMock.target + trim0x(permit1), innerPermitCalldata, { value: 1n });
+        await expect(tx).to.emit(permitAndCallMock, 'MsgValue').withArgs(1n);
```

Comment:
I believe we should make sure there are no duplicate events. Remember, bad thing happened to `multicall` with `ETH` deposit?